fog/view/connection/src/fog_view_router_client.rs

```diff
@@ -1,69 +1,236 @@
 use mc_util_grpc::ConnectionUriGrpcioChannel;
+use mc_util_serial::DecodeError;
+use sha2::Sha512;
 use std::sync::Arc;
 
 /// A high-level object mediating requests to the fog view router service
 pub struct FogViewRouterGrpcClient {
-    /// The fog view router grpc client
-    fog_view_router_client: FogViewRouterApiClient,
     /// A logger object
     logger: Logger,
+
+    /// The AKE state machine object, if one is available.
+    attest_cipher: Option<Ready<Aes256Gcm>>,
+
+    _fog_view_router_client: FogViewRouterApiClient,
+
+    /// Sends requests to the fog view router
+    request_sender: ClientDuplexSender<FogViewRouterRequest>,
+
+    /// Receives responses from the fog view router
+    response_receiver: ClientDuplexReceiver<FogViewRouterResponse>,
+
+    uri: FogViewRouterUri,
+
+    /// An object which can verify a fog node's provided IAS report
+    verifier: Verifier,
 }
 
 impl FogViewRouterGrpcClient {
-    /// Creates a new fog view router grpc client
+    /// Creates a new fog view router grpc client and opens a streaming
+    /// connection to the fog view router service.
     ///
     /// Arguments:
     /// * uri: The Uri to connect to
+    /// * verifier: The attestation verifier
     /// * env: A grpc environment (thread pool) to use for this connection
     /// * logger: For logging
-    pub fn new(uri: FogViewRouterUri, env: Arc<Environment>, logger: Logger) -> Self {
+    pub fn new(
+        uri: FogViewRouterUri,
+        verifier: Verifier,
+        env: Arc<Environment>,
+        logger: Logger,
+    ) -> Self {
         let logger = logger.new(o!("mc.fog.view.router.uri" => uri.to_string()));
 
         let ch = ChannelBuilder::default_channel_builder(env).connect_to_uri(&uri, &logger);
-
         let fog_view_router_client = FogViewRouterApiClient::new(ch);
+        let (request_sender, response_receiver) = fog_view_router_client
+            .request()
+            .expect("Could not retrieve grpc sender and receiver.");
 
         Self {
-            fog_view_router_client,
             logger,
+            attest_cipher: None,
+            _fog_view_router_client: fog_view_router_client,
+            request_sender,
+            response_receiver,
+            uri,
+            verifier,
         }
     }
 
-    /// Makes streaming requests to the fog view router service.
-    pub async fn request(&self) -> Result<(), grpcio::Error> {
-        let (mut sink, mut receiver) = self.fog_view_router_client.request()?;
-        let attested_message = Message::new();
+    fn is_attested(&self) -> bool {
+        self.attest_cipher.is_some()
+    }
+
+    async fn attest(&mut self) -> Result<VerificationReport, Error> {
+        // If we have an existing attestation, nuke it.
+        self.deattest();
+
+        let mut csprng = McRng::default();
+
+        let initiator = Start::new(
+            self.uri
+                .responder_id()
+                .map_err(|_| Error::Attestation)?
```

Comment:
I strongly recommend implementing the appropriate `From<FooError> for Error` traits and not doing this error-swallowing `map_err` stuff. Debugging attestation issues is pretty complex as-is, it's going to rely entirely on intuition when the answer is just "something went wrong in attest()"...

Reply:
Done.